cmd/soroban-rpc/internal/network/backlogQ.go

```diff
@@ -0,0 +1,107 @@
+package network
+
+import (
+	"context"
+	"net/http"
+	"sync/atomic"
+
+	"github.com/creachadair/jrpc2"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/stellar/go/support/errors"
+	"github.com/stellar/go/support/log"
+)
+
+type backlogQLimiter struct {
+	limit        uint64
+	pending      uint64
+	gauge        prometheus.Gauge
+	limitReached uint64
+	logger       *log.Entry
+}
+
+type backlogHTTPQLimiter struct {
+	httpDownstreamHandler http.Handler
+	backlogQLimiter
+}
+
+func MakeHTTPBacklogQueueLimiter(downstream http.Handler, gauge prometheus.Gauge, limit uint64, logger *log.Entry) *backlogHTTPQLimiter {
+	return &backlogHTTPQLimiter{
+		httpDownstreamHandler: downstream,
+		backlogQLimiter: backlogQLimiter{
+			limit:  limit,
+			gauge:  gauge,
+			logger: logger,
+		},
+	}
+}
+
+type backlogJrpcQLimiter struct {
+	jrpcDownstreamHandler jrpc2.Handler
+	backlogQLimiter
+}
+
+func MakeJrpcBacklogQueueLimiter(downstream jrpc2.Handler, gauge prometheus.Gauge, limit uint64, logger *log.Entry) *backlogJrpcQLimiter {
+	return &backlogJrpcQLimiter{
+		jrpcDownstreamHandler: downstream,
+		backlogQLimiter: backlogQLimiter{
+			limit:  limit,
+			gauge:  gauge,
+			logger: logger,
+		},
+	}
+}
+
+func (q *backlogHTTPQLimiter) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	if newPending := atomic.AddUint64(&q.pending, 1); newPending > q.limit {
```

Comment:
just for consideration, allowing `0` for limits which would disable limit checking? as way for deployments to disable this in some cases..but not sure exactly when that would come up..

Reply:
yes, exactly. Explicitly setting the value to 0 would allow us to disable the endpoint.
( the default is non-zero, so the user would need to explicitly have it set to zero in order to disable the endpoint )

Reply:
In the next pr, I'll be adding a pass-through, which would allow the end user to completely skip the limiters.